<!-- Projeto Desenvolvido na Data Science Academy - www.datascienceacademy.com.br -->
# <font color='blue'>Data Science Academy</font>
## <font color='blue'>Matemática e Estatística Aplicada Para Data Science, Machine Learning e IA</font>
## <font color='blue'>Lista de Exercícios 5</font>
## <font color='blue'>Calculando a Similaridade Entre Vetores de Dados</font>

### Instalando e Carregando Pacotes

In [1]:
# Para atualizar um pacote, execute o comando abaixo no terminal ou prompt de comando:
# pip install -U nome_pacote

# Para instalar a versão exata de um pacote, execute o comando abaixo no terminal ou prompt de comando:
# !pip install nome_pacote==versão_desejada

# Depois de instalar ou atualizar o pacote, reinicie o jupyter notebook.

# Instala o pacote watermark. 
# Esse pacote é usado para gravar as versões de outros pacotes usados neste jupyter notebook.
!pip install -q -U watermark

In [2]:
# Versões dos pacotes usados neste jupyter notebook
%reload_ext watermark
%watermark -a "Data Science Academy" 

Author: Data Science Academy



**Resolva os exercícios abaixo usando Linguagem Python. Faça pesquisa complementar se necessário.**

### Exercício 1: Cálculo de Distância Euclidiana

Descrição: Escreva uma função em Python que receba dois vetores (listas de números) de mesma dimensão e retorne a distância euclidiana entre eles.

In [3]:
# Função
def distancia_euclidiana(vetor1, vetor2):
    
    """
    Calcula a distância euclidiana entre dois vetores.

    :param vetor1: Lista de números representando o primeiro vetor.
    :param vetor2: Lista de números representando o segundo vetor.
    :return: A distância euclidiana entre os dois vetores.
    """

    # Calcula a soma das diferenças quadradas entre os elementos correspondentes dos vetores
    soma_diferencas_quadradas = sum((a - b) ** 2 for a, b in zip(vetor1, vetor2))

    # Retorna a raiz quadrada da soma
    return soma_diferencas_quadradas ** 0.5

In [4]:
# Vetores de exemplo
vetor_a = [1, 2, 3]
vetor_b = [4, 5, 6]

In [5]:
# Executa a função
distancia = distancia_euclidiana(vetor_a, vetor_b)
print(distancia)

5.196152422706632


O valor 5.196 representa a distância euclidiana entre os dois vetores fornecidos. A distância euclidiana, também conhecida como norma L2 ou norma euclidiana, é uma medida de distância no espaço euclidiano, que é o espaço geométrico mais comum.

Para interpretar esse valor:

**Contexto Geométrico**: Imagine que cada vetor representa um ponto em um espaço tridimensional (pois cada vetor tem três componentes). O valor 5.196 é a distância "em linha reta" entre esses dois pontos. Se você pudesse traçar uma linha entre os pontos representados pelos vetores [1, 2, 3] e [4, 5, 6], essa linha teria um comprimento de aproximadamente 5.196 unidades.

**Similaridade**: Em algumas aplicações, como aprendizado de máquina e estatística, a distância euclidiana pode ser usada para avaliar a similaridade entre dois conjuntos de dados. Menores distâncias indicam maior similaridade, enquanto maiores distâncias indicam menor similaridade.

**Interpretação Numérica**: O número 5.196, em si, é uma medida absoluta e sua interpretação pode depender do contexto. Por exemplo, se os vetores representam coordenadas em metros, então a distância entre os pontos é de 5.196 metros. Se não houver uma unidade específica ou um contexto, este valor é apenas uma medida de "quão longe" os pontos estão um do outro em termos matemáticos.

### Exercício 2: Distância Euclidiana com Validação de Dados

Descrição: Modifique a função do Exercício 1 para validar se os dois vetores têm o mesmo tamanho. Se não tiverem, a função deve retornar um erro.

In [6]:
# Função
def distancia_euclidiana_com_validacao(vetor1, vetor2):
    
    """
    Calcula a distância euclidiana entre dois vetores com validação de tamanho.

    :param vetor1: Lista de números representando o primeiro vetor.
    :param vetor2: Lista de números representando o segundo vetor.
    :return: A distância euclidiana entre os dois vetores.
    :raises: ValueError se os vetores têm tamanhos diferentes.
    """
    
    # Verifica se os vetores têm o mesmo tamanho
    if len(vetor1) != len(vetor2):
        raise ValueError("Os vetores devem ter a mesma dimensão para calcular a distância.")

    # Calcula a soma das diferenças quadradas entre os elementos correspondentes dos vetores
    soma_diferencas_quadradas = sum((a - b) ** 2 for a, b in zip(vetor1, vetor2))

    # Retorna a raiz quadrada da soma
    return soma_diferencas_quadradas ** 0.5

In [7]:
# Exemplo de uso da função com vetores de tamanhos iguais

vetor_a = [1, 2, 3]
vetor_b = [4, 5, 6]

try:
    distancia = distancia_euclidiana_com_validacao(vetor_a, vetor_b)
    print(f"A distância é: {distancia}")
except ValueError as e:
    print(f"Erro: {e}")

A distância é: 5.196152422706632


In [8]:
# Exemplo com vetores de tamanhos diferentes

vetor_c = [1, 2]
vetor_d = [3, 4, 5]

try:
    distancia = distancia_euclidiana_com_validacao(vetor_c, vetor_d)
    print(f"A distância é: {distancia}")
except ValueError as e:
    print(f"Erro: {e}")

Erro: Os vetores devem ter a mesma dimensão para calcular a distância.


### Exercício 3: Distância Euclidiana em N-Dimensões

Descrição: Generalize a função do Exercício 1 para que ela possa calcular a distância euclidiana entre dois vetores em um espaço de qualquer número de dimensões.

In [9]:
# Função
def distancia_euclidiana_nd(vetor1, vetor2):
    
    """
    Calcula a distância euclidiana entre dois vetores em um espaço de qualquer número de dimensões.

    :param vetor1: Lista de números representando o primeiro vetor.
    :param vetor2: Lista de números representando o segundo vetor.
    :return: A distância euclidiana entre os dois vetores.
    :raises: ValueError se os vetores têm tamanhos diferentes.
    """
    
    # Verifica se os vetores têm o mesmo tamanho
    if len(vetor1) != len(vetor2):
        raise ValueError("Os vetores devem ter a mesma dimensão para calcular a distância.")

    # Calcula a soma das diferenças quadradas entre os elementos correspondentes dos vetores
    soma_diferencas_quadradas = sum((a - b) ** 2 for a, b in zip(vetor1, vetor2))

    # Retorna a raiz quadrada da soma
    return soma_diferencas_quadradas ** 0.5

In [10]:
# Usando a função

vetor_a = [1, 2, 3, 4]
vetor_b = [5, 6, 7, 8]

distancia = distancia_euclidiana_nd(vetor_a, vetor_b)
print(distancia)

8.0


In [11]:
# Usando a função

vetor_a = [0.1, 200, 3/2, 4.9, 10]
vetor_b = [1.2, 1/8, 0.78, 0.003, 20]

distancia = distancia_euclidiana_nd(vetor_a, vetor_b)
print(distancia)

200.18922207251816


In [12]:
# Exemplo com vetores de tamanhos diferentes

vetor_a = [0.1, 200, 3/2, 4.9]
vetor_b = [1.2, 1/8, 0.78, 0.003, 20]

try:
    distancia = distancia_euclidiana_nd(vetor_a, vetor_b)
    print(f"A distância é: {distancia}")
except ValueError as e:
    print(f"Erro: {e}")

Erro: Os vetores devem ter a mesma dimensão para calcular a distância.


### Exercício 4: Comparando Distâncias

Descrição: Escreva um script que receba três vetores de mesma dimensão. Calcule a distância euclidiana entre o primeiro e os outros dois vetores. Imprima qual dos dois vetores está mais próximo do primeiro, baseado na distância euclidiana.

In [13]:
# Função
def distancia_euclidiana(vetor1, vetor2):
    
    """
    Calcula a distância euclidiana entre dois vetores.

    :param vetor1: Lista de números representando o primeiro vetor.
    :param vetor2: Lista de números representando o segundo vetor.
    :return: A distância euclidiana entre os dois vetores.
    """
    
    return sum((a - b) ** 2 for a, b in zip(vetor1, vetor2)) ** 0.5

In [14]:
# Função
def comparar_distancias(vetor1, vetor2, vetor3):
    
    """
    Compara as distâncias euclidianas entre o primeiro vetor e outros dois vetores.

    :param vetor1: Lista de números representando o primeiro vetor.
    :param vetor2: Lista de números representando o segundo vetor.
    :param vetor3: Lista de números representando o terceiro vetor.
    :return: Imprime qual dos dois vetores (vetor2 ou vetor3) está mais próximo do vetor1.
    """
    
    # Calcula as distâncias
    distancia1 = distancia_euclidiana(vetor1, vetor2)
    distancia2 = distancia_euclidiana(vetor1, vetor3)

    # Compara as distâncias e imprime o resultado
    # Menores distâncias indicam maior similaridade, enquanto maiores distâncias indicam menor similaridade
    if distancia1 < distancia2:
        return "O segundo vetor está mais próximo do primeiro."
    elif distancia2 < distancia1:
        return "O terceiro vetor está mais próximo do primeiro."
    else:
        return "Os dois vetores estão à mesma distância do primeiro."

In [15]:
# Vetores de exemplo
vetor_a = [1, 2, 3]
vetor_b = [5.0, 0.7, 1/6]
vetor_c = [1.4, 0.33, 2.2]

In [16]:
# Aplica a função
resultado = comparar_distancias(vetor_a, vetor_b, vetor_c)
print(resultado)

O terceiro vetor está mais próximo do primeiro.


### Exercício 5: Aplicação em Data Science

Descrição: Dada uma matriz de dados (lista de listas, onde cada sublista é um vetor de características) e um vetor de referência, escreva uma função que calcule a distância euclidiana de cada vetor na matriz até o vetor de referência. A função deve retornar uma lista de distâncias.

In [17]:
# Função
def calcular_distancias(matriz, vetor_referencia):
    
    """
    Calcula a distância euclidiana de cada vetor em uma matriz até um vetor de referência.

    :param matriz: Lista de listas, onde cada sublista é um vetor de características.
    :param vetor_referencia: Lista de números representando o vetor de referência.
    :return: Lista de distâncias euclidianas de cada vetor na matriz até o vetor de referência.
    """
    
    distancias = []
    
    for vetor in matriz:
        distancia = distancia_euclidiana(vetor, vetor_referencia)
        distancias.append(distancia)
        
    return distancias

In [18]:
# Matriz de exemplo
matriz_dados = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]

In [19]:
# Vetor de referência
vetor_ref = [1.2, 3.4, 0.998]

In [20]:
# Executa a função
distancias = calcular_distancias(matriz_dados, vetor_ref)
print(distancias)

[2.4511230079292226, 5.9514707425980005, 10.901009311068403]


In [21]:
%reload_ext watermark
%watermark -a "Data Science Academy"

Author: Data Science Academy



In [22]:
#%watermark -v -m

In [23]:
#%watermark --iversions

# Fim